google/cloud/internal/rest_completion_queue_impl.h

```diff
@@ -0,0 +1,254 @@
+      return f;
+    }
+
+    void Shutdown() {
+      std::unique_lock<std::mutex> const lk(mu_);
+      shutdown_ = true;
+      cv_.notify_all();
+    }
+
+    void Service() {
+      while (true) {
+        std::unique_lock<std::mutex> lk(mu_);
+        auto const ne = NextExpiration();
+        cv_.wait_until(
+            lk, ne, [this, ne] { return shutdown_ || ne >= NextExpiration(); });
+        if (shutdown_) return NotifyShutdown(std::move(lk));
+        ExpireTimers(std::move(lk), Clock::now());
+      }
+    }
+
+    std::int64_t timer_counter() const { return timer_counter_.load(); }
+
+   private:
+    void ExpireTimers(std::unique_lock<std::mutex> lk, TimePoint tp) {
+      auto p = std::move(timers_.begin()->second);
+      timers_.erase(timers_.begin());
+      auto const has_expired_timers =
+          !timers_.empty() && timers_.begin()->first < tp;
+      lk.unlock();
+      // We do not want to expire all the timers in this thread, as this would
+      // serialize the expiration.  Instead, we pick another thread to continue
+      // expiring timers.
+      if (has_expired_timers) cv_.notify_one();
+      p.set_value(tp);
+      ++timer_counter_;
+    }
+
+    void NotifyShutdown(std::unique_lock<std::mutex> lk) {
+      auto timers = std::move(timers_);
+      lk.unlock();
+      for (auto& kv : timers) {
+        kv.second.set_value(
+            Status(StatusCode::kCancelled, "timer queue shutdown"));
+      }
+    }
+
+    TimePoint NextExpiration() {
+      auto constexpr kMaxSleep = std::chrono::hours(1);
+      if (timers_.empty()) return Clock::now() + kMaxSleep;
+      return timers_.begin()->first;
+    }
+
+    using PromiseType = promise<StatusOr<TimePoint>>;
+    std::mutex mu_;
+    std::condition_variable cv_;
+    std::atomic<std::int64_t> timer_counter_{0};
+    std::multimap<TimePoint, PromiseType> timers_;
+    bool shutdown_ = false;
+  };
+
+  /// Start an operation with the lock already held.
+  void StartOperation(std::unique_lock<std::mutex> lk,
+                      std::shared_ptr<AsyncRestOperation> op,
+                      absl::FunctionRef<void(void*)> start);
+
+  /// Return the asynchronous operation associated with @p tag.
+  std::shared_ptr<AsyncRestOperation> FindOperation(void* tag);
+
+  /// Unregister @p tag from pending operations.
+  void ForgetOperation(void* tag);
+
+  void RunStart();
+  void RunStop();
+
+  void DrainRunAsyncLoop();
+  void DrainRunAsyncOnIdle();
+  void WakeUpRunAsyncThread(std::unique_lock<std::mutex> lk);
+
+  class WakeUpRunAsyncLoop;
+  class WakeUpRunAsyncOnIdle;
+
+  std::mutex mu_;
+  grpc::CompletionQueue grpc_cq_;
+  RestCompletionQueue cq_;
+  std::size_t thread_pool_size_ = 0;
+  std::size_t run_async_pool_size_ = 0;
+  std::deque<std::unique_ptr<internal::RunAsyncBase>> run_async_queue_;
+  bool shutdown_{false};  // GUARDED_BY(mu_)
+  std::unordered_map<void*, std::shared_ptr<AsyncRestOperation>>
+      pending_ops_;  // GUARDED_BY(mu_)
```

Comment:
This seems to be where much of the `void*` stuff is originating.  Why, for example, can't it be a set of (shared pointers to) operations?